src/realtime.py

```diff
@@ -8,91 +8,117 @@
-                log.info("converting record", record=record)
-                converted = pydarnio.BorealisConvert.\
-                    _BorealisConvert__convert_rawacf_record(0, (record, rawacf_data[record]), "")
-            except pydarnio.borealis_exceptions.BorealisConvert2RawacfError:
-                log.info("error converting")
-                continue
-
-            for rec in converted:
-                fit_data = fitacf._fit(rec)
-                tmp = fit_data.copy()
-
-                # Can't jsonify numpy, so we convert to native types for realtime purposes.
-                for k, v in fit_data.items():
-                    if hasattr(v, 'dtype'):
-                        if isinstance(v, np.ndarray):
-                            tmp[k] = v.tolist()
-                        else:
-                            tmp[k] = v.item()
-
-                q.put(tmp)
-
-    def handle_remote_connection():
-        """
-        Compresses and serializes the data to send to the server.
-        """
-
-        while True:
-            data_dict = q.get()
-            serialized = json.dumps(data_dict)
-            compressed = zlib.compress(serialized.encode('utf-8'))
-            realtime_socket.send(compressed)
-
-    threads = [threading.Thread(target=get_temp_file_from_datawrite),
-               threading.Thread(target=handle_remote_connection)]
-
-    for thread in threads:
-        thread.daemon = True
-        thread.start()
-
-    for thread in threads:
-        thread.join()
+    # Socket for receiving data from data_write
+    data_write_socket = so.create_sockets([options.rt_to_dw_identity], options.router_address)[0]
 
+    # Socket for serving data over the web
+    publish_socket = context.socket(zmq.PUB)
+    publish_socket.setsockopt(zmq.LINGER, 0)  # milliseconds to wait for message to send when closing socket
 
-if __name__ == '__main__':
-    from utils import log_config
+    try:
+        publish_socket.bind(options.realtime_address)
+    except zmq.ZMQError as e:  # Raised if the address is invalid (e.g. if device doesn't exist)
+        log.critical("REALTIME CRASHED", error=e)
+        log.exception("REALTIME CRASHED", exception=e)
+        data_write_socket.close()
+        publish_socket.close()
 
-    log = log_config.log()
-    log.info(f"REALTIME BOOTED")
     try:
-        main()
+        realtime_server(data_write_socket, publish_socket)
         log.info(f"REALTIME EXITED")
+    except KeyboardInterrupt:
+        log.critical("REALTIME INTERRUPTED")
     except Exception as main_exception:
         log.critical("REALTIME CRASHED", error=main_exception)
         log.exception("REALTIME CRASHED", exception=main_exception)
+
+else:
+    from .utils import socket_operations as so
+    from .utils.options import Options
+
+    caller = Path(inspect.stack()[-1].filename)
+    module_name = caller.name.split('.')[0]
+    log = structlog.getLogger(module_name)
```

Comment:
What is this else block for?  Is it for importing the realtime server?

Reply:
Yes this is required for the `realtime_sim.py` script. Without this, the imports don't work properly, and neither does logging.